context.go

```diff
@@ -453,10 +453,32 @@ func (ctx Context) App(name string) (any, error) {
 	return modVal, nil
 }
 
+// AppStrict is like App, but it returns an error if the app
+// has not been configured. This is useful when the app is
+// required and its absence is a configuration error.
+func (ctx Context) AppStrict(name string) (any, error) {
+	if app, ok := ctx.cfg.apps[name]; ok {
+		return app, nil
+	}
+	appRaw := ctx.cfg.AppsRaw[name]
+	if appRaw == nil {
+		return nil, fmt.Errorf("app module %s is not configured", name)
+	}
+	return ctx.App(name)
+}
+
 // AppIfConfigured returns an app by its name if it has been
-// configured. Can be called instead of App() to avoid
-// instantiating an empty app when that's not desirable. If
-// the app has not been loaded, nil is returned.
+// configured and provisioned. Can be called instead of App()
+// to avoid instantiating an empty app when that's not desirable.
+// If the app has not been loaded, nil is returned.
+//
+// Since this method does not return the app if it has not been
+// provisioned, this could be misleading if the provisioning order
+// caused the app to not be provisioned yet. This could return
+// nil even if the app is configured, but not yet provisioned.
+// In that case, the caller should probably use AppStrict() instead
+// to ensure the app is provisioned, and no app will be instantiated
+// if it is not configured.
```

Comment:
So the more I read this, the more I think we need to just fix this: if the app has been configured, load and provision it if needed. Otherwise, don't. This method is called `AppIfConfigured` not `AppIfProvisionedAlready`...

What do you think if we try to do that?

Reply:
Assuming you had a particular need when you changed it in https://github.com/caddyserver/caddy/commit/0e2c7e1d35b287fc0e56d6db2951f791e09b5a37, are those semantics not needed anymore?

Reply:
I don't think either change is quite right though...

The earlier version would provision an app that maybe shouldn't be provisioned (I forgot about the nondeterminstic order). The change makes it never provision the app, but also disregards the loading order.

AppIfConfigured should provision the app if, but ONLY if, it _will_ be provisioned.

Reply:
I don't think I follow. So what do you think should be done? Just rename AppStrict to AppIfConfigured (and drop the current one & functionality)?

Why did the difference matter for the cert cache commit again? Can you give an example of when it would have been wrong?

Reply:
I'm a dummy, `AppStrict()` already does exactly what I'm describing, lol.

So yeah, let's replace AppIfConfigured with AppStrict. This does add an `error` return to the signature but we had it originally and I don't think there's anyone using this yet? (maybe 1 or 2? I apologize if that breaks anyone, but it's a very very easy patch, and I think this is an important fix).